### ISSUE TITLE
It is not possible to create Schema with array of referenced items

### ISSUE THREAD

**Issue by KrystianRos** [eclipse/microprofile-open-api#425]
Versions: 1.0.1, 1.1.2, 2.0-MR1
I'm using the following annotation for schema:
`@Schema(required = true, description = "Some description...", type = SchemaType.ARRAY, ref = "name1")`
What I'm getting is:
```
variableName:
          $ref: '#/components/schemas/name1'
```
What I expect to get:
```
variableName:
          type: array
          description: Some description...
          items:
                    $ref: '#/components/schemas/name1'
```
It looks like ref is overriding all of data in @Schema object. I have tried a lot of modifications for mentioned annotation, but only pseudo-success is when I have add an 'implementation' element pointing to class- but that's not satisfying solution.
I'm sorry for not pasting original content, but it's protected by contract.

**Comment by EricWittmann** [comment]
Yes that is the contract for the `ref` property.  That property is mutually exclusive with all other properties on `@Schema`.

**Comment by EricWittmann** [comment]
One thing you could try is creating a dummy class/interface like this:

```java
interface NamedList extends List<Named>
```
Then you can use that as the implementation class for the annotation.

**Comment by EricWittmann** [comment]
@MikeEdgar @arthurdm any thoughts on other workarounds?  I feel like I'm missing an obvious approach, but I can't think of anything.

**Comment by MikeEdgar** [comment]
What about the following where `NamedBean` is the class containing the schema you previously referenced as `"name1"` If you wanted to ensure the `$ref` is a particular value, you should annotate the `NamedBean` class with `@Schema(name = "name1")` as well.

```java
@Schema(
    required = true,
    description = "Some description...", 
    type = SchemaType.ARRAY,
    implementation = NamedBean.class)
```

**Comment by KrystianRos** [comment]
Hi @MikeEdgar, when I'm using 'implementation' in combination with SchemaType.ARRAY, then in result final document I'm getting a full description of schemas which are over all fields inside NamedBean, not $ref field (if I understand correctly. what you mean).

**Comment by MikeEdgar** [comment]
@KrystianRos - did you try @EricWittmann's approach of using an interface?

**Comment by 0xqab** [comment]
Sitting on the same boat as @KrystianRos.
I tried the approach with the interface, but at least on Payara Micro the NamedList interface is not listed in the components sections, hence the `"$ref": "#/components/schemas/NamedList"` points nowhere. I assume I have to create a class instead, which is really ugly because of the amount of functions which have to be implemented. This is not a feasable way for a huge API with lot's of lists.

**Comment by MikeEdgar** [comment]
@KrystianRos, @0xqab - I would suggest you try either 

(1) implementing an `OASModelReader` and having it produce the correct array `Schema` in `components` which can then be referenced by name using `ref` in `@Schema` annotations throughout your code or
(2) Specify the same schema in components in a static YAML or JSON file which can also be referenced with `ref`.

@0xqab mentioned the use of Payara, which implementation are you using @KrystianRos ? I expect the type = ARRAY with `implementation = MyClass.class` to work with runtimes using SmallRye.

**Comment by MikeEdgar** [comment]
This issue was discussed in the 2023-10-02 meetup call. One proposed solution would be to allow for the `@Schema` annotation to be placed on `TYPE_USE` elements. This would split the `items` schema from the `array` schema into separate uses of the annotation.

With that enhancement, developers could write something like the code below and get the expected result.

```java
@Schema(required = true, description = "Some description...")
List<@Schema(ref = "name") MyNameBean> variableName;
```

```yaml
variableName:
  type: array
  description: Some description...
  items:
    $ref: '#/components/schemas/name1'
```

cc: @tjquinno @Azquelt

**Comment by Azquelt** [comment]
We noted a few shortcomings for Option 1:

- If the field is an array, I don't think there's a place to put the annotation for the item schema:
  ```
  @Schema(required = true, description = "Some description...")
  MyNameBean[] variableName;
  ```
- If the array maps to a single Java object, rather than a collection type, there's also no place to put the item schema:
  ```
  @Schema(required = true, description = "Some description...", type = ARRAY)
  MyNameBeansHolder variableName;
  ```
- If the schema is declared within `@Components(schemas = { ... })`, there's no way to declare the item schema
  - I guess there's not really a good way around this. The current solution for `allOf` requires another object anyway, it can't encompass the schemas within the `@Components` annotation.

In these three scenarios, Option 3 would provide a way to declare the item schema, at the expense of requiring an extra object to host the `@Schema` annotation.

**Comment by MikeEdgar** [comment]
> * If the field is an array, I don't think there's a place to put the annotation for the item schema:
>       ```
>       @Schema(required = true, description = "Some description...")
>       MyNameBean[] variableName;
>       ```

It looks odd, but this is actually possible:
```java
MyNameBean @Schema(required = true, description = "Some description...")[] variableName;
```

> * If the array maps to a single Java object, rather than a collection type, there's also no place to put the item schema

I agree, this would not be supported without some other class to reference, either using `implementation = ItemBean[].class` or a new `items = ItemBean.class`. The first one using an array for `implementation` still has all the limitations of this issue, so it's not really an option for many cases.

> * If the schema is declared within `@Components(schemas = { ... })`, there's no way to declare the item schema

Agreed, possibly this placement of the annotation is the strongest argument for adding `Class<?> items()` to `@Schema`. Even though there would technically be two ways to describe array items, I still think TYPE_USE provides greater functionality beyond just arrays - e.g. `additionalProperties` and certainly others we're not thinking about.

**Comment by MikeEdgar** [comment]
Annotating the array type as suggested above may not actually be what is wanted. I need to look further into how the annotated is represented.

**Comment by MikeEdgar** [comment]
It appears that when annotating with `TYPE_USE`, the annotation directly on the array element type is the one that would be considered the `items` schema. The same annotation also appears on the field in the Java class, so in cases like this the field annotation would need to be discarded by a scanner implementation. The annotation on the array type `[]` would then be the _outer_ schema and should support arbitrary levels of nesting.

It's probably not what people are used to seeing since it would be reversed from the typical way of annotating fields, parameters, etc. However, it technically would allow for the schema to be expressed as needed when using arrays.

```java
@Schema(description = "Item description") String 
  @Schema(description = "Array description")[] variableName;
```
Resulting schema:
```json
"variableName": {
  "type": "array",
  "description": "Array description",
  "items": {
    "type": "string",
    "description": "Item description"
  }
}
```

**Comment by Azquelt** [comment]
We would want to make sure that we don't change the meaning of this:
```java
@Schema(description = "Item/array description?") String[] variableName;
```

Today, this would refer to the field type (i.e. the array) whereas in the above example, it refers to the item type.

**Comment by Azquelt** [comment]
From the call: we could look at how bean validation handles type annotations on array fields like this.

**Comment by mklueh** [comment]
I was stumbling over this issue as well, trying to integrate OpenAPI in my application, and use **oneOf** / **anyOf** together with type ARRAY in my response object in the most (in my opinion) logical and intuitive way:

```
@Data
@Builder
@AllArgsConstructor
public class SomeResponseObject {

    private long page;

    private long pages;

    @Schema(type = ARRAY, oneOf = {SomeItem.class, SomeOtherItem.class})
    private List<BaseClass> items;
}
```

This however will lead to

```
 "SomeResponseObject" : {
        "type" : "object",
        "properties" : {
          "page" : {
            "format" : "int64",
            "type" : "integer"
          },
          "pages" : {
            "format" : "int64",
            "type" : "integer"
          },
          "items" : {
            "type" : "array",
            "items" : {
              "$ref" : "#/components/schemas/BaseClass"
            },
             "oneOf" : [ {
              "$ref" : "#/components/schemas/SomeItem"
            }, {
              "$ref" : "#/components/schemas/SomeOtherItem"
            } ]
          }
        }
      },
```

which generates his typescript object
```
    SomeResponseObject: {
      /** Format: int64 */
      page?: number;
      /** Format: int64 */
      pages?: number;
      items?: components["schemas"]["SomeItem"] | components["schemas"]["SomeOtherItem"];
    };
```

without any array type.

This is the current vs the expected result discussed here https://github.com/drwpow/openapi-typescript/issues/1421#issuecomment-1792334575:

![image](https://github.com/eclipse/microprofile-open-api/assets/1553757/2b93a582-14f7-437a-b256-47aa98b5f6ef)


**While it does not do what I expected, I'm wondering what use case the actual result has and why it can't do what it should instead technically.**

I mean, when do I want to annotate a list with **type array** and **anyOf** / **oneOf** and really want something different as a result than specifying a type **array** that is one of the n types?

**Comment by Azquelt** [comment]
@mklueh Thank you for your detailed breakdown of the problems you've had trying to the types right for an array. I'm sorry I've been really busy and still haven't had time to sit down and go through what you've written but I hope to get to it later this week.

**Comment by Azquelt** [comment]
@mklueh's suggestion amounts to saying that when `@Schema` is placed on an array element, some of the properties should apply to the array itself, while others should apply to the array items.

I'm not sure this is a good idea because while there are some fields that might be unambiguous, there are others which are not. In particular:
- `type` has to refer to the array itself, in the example we have `type=ARRAY`, but defining the type of the item is also something you may want to do
- `nullable` could refer to the array itself being null, or to allowing `null` as an item in the array
- `extensions` could refer to the array itself, or to the type of the items
- all of the documentation fields (`title`, `description`, `example`, `externalDocs`) could refer to either type
- I would argue that `allOf`, `oneOf` and `anyOf` could be valid for the array itself - you could want to say "this property can be either an array or a string"

However, I note that this is also a problem we have for objects which we solved with the `properties` field, which takes an array of `@PropertySchema` (which itself has most of the same fields as `@Schema`).

Would it make sense to have an `items` or `itemSchema` field which holds a `@PropertySchema` (or another similar annotation)?

For @mklueh's example, you could have
```java
    @Schema(type = ARRAY, items = @PropertySchema(oneOf = {BaseClass.class, SomeItem.class, SomeOtherItem.class}))
    private List<BaseClass> items;
```

For the original example where you want to refer to a component schema for the item type you could have
```java
    @Schema(required = true, description = "Some description...", items = @PropertySchema(ref="name))
    List<MyNameBean> variableName;
```

As-is, I don't think we can quite just use `@PropertySchema` because the `name` field is currently required. We could either add a default value for `name()` (which I _think_ would be backwards compatible), or we could add a new similar annotation `@ItemSchema`.

**Comment by mklueh** [comment]
@Azquelt thanks for your detailed response.

> @mklueh's suggestion amounts to saying that when `@Schema` is placed on an array element, some of the properties should apply to the array itself, while others should apply to the array items.

Now I see the problem. The annotation would always refer to the types of the field it is annotating.

What if Microprofile would automatically declare any inheritors of the used base type of the collection as **allOf** by default? Then no one would even have to bother with this case, except for **more restrictive edge cases**. (ex. only specific child classes should be allowed)
 
> I'm not sure this is a good idea because while there are some fields that might be unambiguous, there are others which are not. In particular:
> 
> * `type` has to refer to the array itself, in the example we have `type=ARRAY`, but defining the type of the item is also something you may want to do

In case you want to refer to the item, wouldn't you then place the annotation on top of the item's class instead?

> * I would argue that `allOf`, `oneOf` and `anyOf` could be valid for the array itself - you could want to say "this property can be either an array or a string"

Just to make sure I understand you correctly: Does that mean you could declare a field:

```
@Schema(type = ARRAY, oneOf = { String.class})
private Collection<BaseClass> items;
```

to tell Microprofile the **Collection type** should be seen as a **String**? 
If so, what sense does it make or would this only be a valid case if the field would be of type **Object**?

And why is the `type = Array` **property even needed**? Wouldn't it be possible to determine the type automatically if the field's type is Collection or Array?



> However, I note that this is also a problem we have for objects which we solved with the `properties` field, which takes an array of `@PropertySchema` (which itself has most of the same fields as `@Schema`).
> 
> Would it make sense to have an `items` or `itemSchema` field that holds a `@PropertySchema` (or another similar annotation)?
> 
> For @mklueh's example, you could have
> 
> ```java
>     @Schema(type = ARRAY, items = @PropertySchema(oneOf = {BaseClass.class, SomeItem.class, SomeOtherItem.class}))
>     private List<BaseClass> items;
> ```

That would also be nice. Or the property should be named 

```java
     @Schema(type = ARRAY,  itemSchema = @ItemSchema(oneOf = {BaseClass.class, SomeItem.class, SomeOtherItem.class}))
     private List<BaseClass> items;
```

to make it a bit clearer

> For the original example where you want to refer to a component schema for the item type you could have
> 
> ```java
>     @Schema(required = true, description = "Some description...", items = @PropertySchema(ref="name))
>     List<MyNameBean> variableName;
> ```
> 
> As-is, I don't think we can quite just use `@PropertySchema` because the `name` field is currently required. We could either add a default value for `name()` (which I _think_ would be backward compatible), or we could add a new similar annotation `@ItemSchema`.

Also nice, but not sure which of both variants would be more straightforward from a user's perspective...I'd generally advocate against yet another annotation, but thinking about it, maybe `@ItemSchema` would be even clearer considering that `@Schema` is only for the field type.

`@FieldSchema` and `@ItemSchema` would be perfectly distinguishable but would break backward compatibility ofc.

**Comment by MikeEdgar** [comment]
I agree that using `@PropertySchema` with `items` seems to be a reasonable solution here. Not as powerful as allowing `TYPE_USE`, but more readable and probably easier to understand.

Adding a default for `name()` should be backward compatible, but we'll probably want to add something to the TCK to require a non-blank value when `@PropertySchema` is being used in `properties`. I think I prefer reusing that annotation to adding `@ItemSchema` that is otherwise identical.

**Comment by Azquelt** [comment]
On the call we broadly agreed that reusing `@PropertySchema` is preferable to adding another nearly identical annotation, even if the naming is not ideal.

I also note that this would provide another way to do #445 if you can't annotate the item class.

**Comment by mklueh** [comment]
Great! Is there a rough timeframe for when a solution might be released?

**Comment by mklueh** [comment]
@Azquelt thanks :)

One thing confuses me when looking at the TCK

![image](https://github.com/eclipse/microprofile-open-api/assets/1553757/aeab76a6-aa5e-443a-9e37-7408b96e0e01)

It seems the `implementation` property applies to the **item type** of the list while the **oneOf, anyOf, allOf** properties apply to the **list type**.

Wouldn't it be better to **deprecate** `implementation` in favor of using `@Schema(type= Array, items = { @PropertySchema( oneOf=Flight.class ) })` to get some more **consistency**?
